### PR TITLE
Update protobuf to 3.7.1 (PR 1): add new files

### PR DESCRIPTION
**Pull request by laszlocsomor**
On behalf of @davido, merging his PR 1: https://github.com/bazelbuild/bazel/commit/026afdb8b14d69d442a32475d146d2b255f0eb08

Original PR description:

> //CC @drigz @laurentlb @meteorcloudy.
>
> Note, that 3.7.1 wasn't tagged yet on the `protobuf` side. I used the code from this commit on 3.7.x branch: > https://github.com/protocolbuffers/protobuf/commit/4b9a5df4e8ba2066794da56598ad2905dc42051e. This was the backport of my fix from master: [1].
>
> [1] https://github.com/protocolbuffers/protobuf/pull/5811

Protobuf update workflow:
https://github.com/bazelbuild/bazel/blob/master/third_party/protobuf/README.md

This PR must be merged directly to
https://bazel.googlesource.com/bazel/+/master.

See https://github.com/bazelbuild/bazel/pull/7681
See https://github.com/bazelbuild/bazel/issues/7611

Change-Id: Ib6edee142ed3cf92806f2b63d119ae11ebe51a6f

### CODE REVIEW

**Comment by laszlocsomor** [comment]
> The bad news is that it appears that one or more commits were authored or co-authored by someone other than the pull request submitter. We need to confirm that all authors are ok with their commits being contributed to this project. Please have them confirm that here in the pull request.

/cc @davido

**Comment by buchgr** [comment]
LGTM

**Comment by davido** [comment]
@laszlocsomor Thanks for uploading. Reviewing this PR now, I see that I erroneously included this file:

third_party/protobuf/3.7.1/0001-Add-dependency-on-error_prone_annotations.patch

Can you please remove it?

**Comment by laszlocsomor** [comment]
Yes, I'll do that.

Sorry I didn't ask earlier, I'm having second thoughts: why does Bazel need a newer protobuf? [Here](https://github.com/bazelbuild/bazel/pull/7681#issuecomment-473186729) you answered:

> Upgrade to the recent protobuf release (3.7.0) is needed to make Bazel build tool chain forward compatible (incompatible options).

Which incompatible change is that?

Jakob asks the same: https://github.com/bazelbuild/bazel/issues/7611#issuecomment-474780758

**Comment by davido** [comment]
>Which incompatible change is that?

//CC @buchgr, @laurentlb  

For one this is: `incompatible_package_name_is_a_function`, see also this issue https://github.com/google/bazel-common/issues/51, for another there is a number of Bazel deprecation/incompatible feature removals, that you can see in `rules_closure` that is used in Gerrit Code Review. Just try to build rules_closure HEAD with most recent Bazel version using `--incompatible-all` flag.

**Comment by laszlocsomor** [comment]
Thanks!

IIUC https://github.com/google/bazel-common/issues/51 concerns [protobuf in google-common](https://github.com/google/bazel-common/blob/f1115e0f777f08c3cdb115526c4e663005bec69b/workspace_defs.bzl#L244). `rules_closure`'s `@com_google_protobuf` is 3.6.1. Both projects need to update the dependency, I understand that.

But Bazel's [customized copy of protobuf 3.6.1](https://github.com/bazelbuild/bazel/tree/master/third_party/protobuf) is only used for building Bazel (?), it's a private dependency of Bazel itself. Do you agree?

Maybe some projects depend on the Bazel source tree (via `git_repository` or `http_archive`) and use `//third_pary/protobuf` from it?

**Comment by davido** [comment]
> But Bazel's customized copy of protobuf 3.6.1 is only used for building Bazel (?), it's a private dependency of Bazel itself. Do you agree?

I'm not sure. How/where `protoc` in Bazel is coming from? See this complication for more details, when protoc bazel version mismatch with protobuf_java version used in downstream project:

https://github.com/bazelbuild/bazel/issues/6852

**Comment by davido** [comment]
@laszlocsomor @buchgr 

When I build gerrit code review with Bazel@HEAD, even with the updated `rules_cosure`  with Protobuf 3.7.0 version (https://github.com/bazelbuild/rules_closure/pull/341/commits), and pass this incompatible option: `--incompatible_disable_deprecated_attr_params` 

I'm getting this failure:

```
  $ bazel build --incompatible_disable_deprecated_attr_params java/com/google/gerrit/server:server
INFO: Invocation ID: b4477d95-d3dc-48a9-8b2b-3dc4a35a5c64
DEBUG: /home/davido/.cache/bazel/_bazel_davido/5c01f4f713b675540b8b424c5c647f63/external/bazel_skylib/lib/versions.bzl:98:7: 
Current Bazel is not a release version, cannot check for compatibility.
DEBUG: /home/davido/.cache/bazel/_bazel_davido/5c01f4f713b675540b8b424c5c647f63/external/bazel_skylib/lib/versions.bzl:99:7: Make sure that you are running at least Bazel 0.22.0.
ERROR: /home/davido/.cache/bazel/_bazel_davido/5c01f4f713b675540b8b424c5c647f63/external/com_google_protobuf/protobuf.bzl:130:19: Traceback (most recent call last):
	File "/home/davido/.cache/bazel/_bazel_davido/5c01f4f713b675540b8b424c5c647f63/external/com_google_protobuf/protobuf.bzl", line 125
		rule(attrs = {"srcs": attr.label_list...()}, <2 more arguments>)
	File "/home/davido/.cache/bazel/_bazel_davido/5c01f4f713b675540b8b424c5c647f63/external/com_google_protobuf/protobuf.bzl", line 130, in rule
		attr.label(cfg = "host", executable = True, sin..., ...)
'single_file' is no longer supported. use allow_single_file instead. You can use --incompatible_disable_deprecated_attr_params=false to temporarily disable this check.
ERROR: /home/davido/.cache/bazel/_bazel_davido/5c01f4f713b675540b8b424c5c647f63/external/io_bazel_rules_closure/java/io/bazel/rules/closure/worker/BUILD:17:1: error loading package '@com_google_protobuf//': Extension file 'protobuf.bzl' has errors and referenced by '@io_bazel_rules_closure//java/io/bazel/rules/closure/worker:worker'
ERROR: Analysis of target '//:gerrit' failed; build aborted: error loading package '@com_google_protobuf//': Extension file 'protobuf.bzl' has errors
INFO: Elapsed time: 1.670s
INFO: 0 processes.
```

I think that this protobuf version is leaked from Bazel's Protobuf fork.

When I, however, update Protobuf in Bazel tree with my series to 3.7.x branch, and re-run the same command, I still see the same error. So, the question is, where the outdated external/com_google_protobuf/protobuf.bzl is coming from?

**Comment by davido** [comment]
Update:

With @dslomov's help, this PR: https://github.com/bazelbuild/rules_closure/pull/341 to upgrade Protobuf version in `rules_closure` was merged. In: [1], I bumped the rules_closure version in gerrit.

Now I merged forward rules_clsorure upgrade to stable-2.16 code base, that added dependency on protoc: [2]. As I said in my previous comment, even if If try to build with custom built Bazel version (with protobuf upgraded to 3.7.1 version) that code base with `--incompatible_disable_deprecated_attr_params`, I'm seeing this failure:

```
$ b10 build --incompatible_disable_deprecated_attr_params java/com/google/gerrit/server:server
INFO: Invocation ID: 694c6b25-f9b6-4dba-b7c1-f8650bc2905f
DEBUG: /home/davido/.cache/bazel/_bazel_davido/5c01f4f713b675540b8b424c5c647f63/external/bazel_skylib/lib/versions.bzl:98:7: 
Current Bazel is not a release version, cannot check for compatibility.
DEBUG: /home/davido/.cache/bazel/_bazel_davido/5c01f4f713b675540b8b424c5c647f63/external/bazel_skylib/lib/versions.bzl:99:7: Make sure that you are running at least Bazel 0.22.0.
ERROR: /home/davido/.cache/bazel/_bazel_davido/5c01f4f713b675540b8b424c5c647f63/external/com_google_protobuf/protobuf.bzl:130:19: Traceback (most recent call last):
	File "/home/davido/.cache/bazel/_bazel_davido/5c01f4f713b675540b8b424c5c647f63/external/com_google_protobuf/protobuf.bzl", line 125
		rule(attrs = {"srcs": attr.label_list...()}, <2 more arguments>)
	File "/home/davido/.cache/bazel/_bazel_davido/5c01f4f713b675540b8b424c5c647f63/external/com_google_protobuf/protobuf.bzl", line 130, in rule
		attr.label(cfg = "host", executable = True, sin..., ...)
'single_file' is no longer supported. use allow_single_file instead. You can use --incompatible_disable_deprecated_attr_params=false to temporarily disable this check.
ERROR: /home/davido/projects/gerrit2/proto/BUILD:1:1: every rule of type proto_library implicitly depends upon the target '@com_google_protobuf//:protoc', but this target could not be found because of: error loading package '@com_google_protobuf//': Extension file 'protobuf.bzl' has errors
ERROR: Analysis of target '//java/com/google/gerrit/server:server' failed; build aborted: error loading package '@com_google_protobuf//': Extension file 'protobuf.bzl' has errors
INFO: Elapsed time: 7.241s
INFO: 0 processes.
```

What am I missing?

[1] https://gerrit-review.googlesource.com/c/gerrit/+/218584
[2] https://gerrit-review.googlesource.com/c/gerrit/+/218586

**Comment by laszlocsomor** [comment]
@davido:

Gerrit's (https://github.com/GerritCodeReview/gerrit/commit/d09ff79720bf472b0ded8faa62af63f9a1b950d9) `@com_google_protobuf` comes from `@io_bazel_rules_closure`'s `closure/repositories.bzl` [1], and Gerrit's `@io_bazel_rules_closure` is defined in the `WORKSPACE` [2].

I cloned https://github.com/bazelbuild/rules_closure/commit/316e6133888bfc39fb860a4f1a31cfcbae485aef, used that as `@io_bazel_rules_closure` in Gerrit (`local_repository`), but `@com_google_protobuf` was still `3.6.1.3` (per bazel query).

I cloned https://github.com/protocolbuffers/protobuf/commit/582743bf40c5d3639a70f98f183914a2c0cd0680 and  updated my `closure/repositories.bzl`'s `com_google_protobuf` and `com_google_protobuf_js` to use that. Building Gerrit now failed because `@zlib` was undefined.

I copied `zlib`'s (and the dependent `net_zlib`'s) definitions from protobuf [3], added them to Gerrit's `WORKSPACE` and could build `//java/com/google/gerrit/server` with `bazel-head` and `--incompatible_disable_deprecated_attr_params`. (I built `bazel-head` from https://github.com/bazelbuild/bazel/commit/e58f1808471517108fe681f42f6760736f8519ee using Bazel 0.23.2, so it still had the old protobuf in `//third_party`)

I don't believe protobuf is leaking from Bazel, because to build `proto_library` rules in a blank workspace I first had to define `com_google_protobuf`, and `protoc` is not embedded in the Bazel binary [4].

**Conclusion**: I think you must update `rules_closure`'s `closure/repositories.bzl`'s definitions of `com_google_protobuf` and `com_google_protobuf_js`.

---

[1] At https://github.com/GerritCodeReview/gerrit/commit/d09ff79720bf472b0ded8faa62af63f9a1b950d9:

    $ ./bazel-head query //external:com_google_protobuf --output=build
    (...)
    # /usr/local/google/home/laszlocsomor/.cache/bazel/_bazel_laszlocsomor/f35dff72f0746c1ac37b42d041d8513e/external/io_bazel_rules_closure/closure/repositories.bzl:677:5
    http_archive(
      name = "com_google_protobuf",
      urls = ["https://mirror.bazel.build/github.com/google/protobuf/archive/v3.6.1.3.tar.gz", "https://github.com/protocolbuffers/protobuf/archive/v3.6.1.3.tar.gz"],
      sha256 = "73fdad358857e120fd0fa19e071a96e15c0f23bb25f85d3f7009abfd4f264a2a",
      strip_prefix = "protobuf-3.6.1.3",
    )

[2] Still at https://github.com/GerritCodeReview/gerrit/commit/d09ff79720bf472b0ded8faa62af63f9a1b950d9:

      $ ./bazel-head query //external:io_bazel_rules_closure --output=build
    (...)
    DEBUG: /usr/local/google/home/laszlocsomor/.cache/bazel/_bazel_laszlocsomor/f35dff72f0746c1ac37b42d041d8513e/external/bazel_skylib/lib/versions.bzl:99:7: Make sure that you are running at least Bazel 0.22.0.
    # /usr/local/google/home/laszlocsomor/stuff/gerrit/WORKSPACE:15:1
    http_archive(
      name = "io_bazel_rules_closure",
      urls = ["https://github.com/bazelbuild/rules_closure/archive/87d24b1df8b62405de8dd059cb604fd9d4b1e395.tar.gz"],
      sha256 = "0e6de40666f2ebb2b30dc0339745a274d9999334a249b05a3b1f46462e489adf",
      strip_prefix = "rules_closure-87d24b1df8b62405de8dd059cb604fd9d4b1e395",
    )

[3] https://github.com/protocolbuffers/protobuf/commit/582743bf40c5d3639a70f98f183914a2c0cd0680 


    $ ~/stuff/gerrit/bazel-head query //external:zlib --output=build
    # /usr/local/google/home/laszlocsomor/stuff/protobuf/WORKSPACE:73:1
    bind(
      name = "zlib",
      actual = "@net_zlib//:zlib",
    )

    $ ~/stuff/gerrit/bazel-head query //external:net_zlib --output=build
    # /usr/local/google/home/laszlocsomor/stuff/protobuf/WORKSPACE:25:1
    http_archive(
      name = "net_zlib",
      urls = ["https://zlib.net/zlib-1.2.11.tar.gz"],
      sha256 = "c3e5e9fdd5004dcb542feda5ee4f0ff0744628baf8ed2dd5d66f8ca1197cb1a1",
      strip_prefix = "zlib-1.2.11",
      build_file = "//:third_party/zlib.BUILD",
    )

[4] See:

    $ unzip -l ./bazel-head | grep proto
     4896  1980-01-01 00:00   embedded_tools/src/main/protobuf/BUILD
     1811  1980-01-01 00:00   embedded_tools/src/main/protobuf/action_cache.proto
     6559  1980-01-01 00:00   embedded_tools/src/main/protobuf/analysis.proto
     1387  1980-01-01 00:00   embedded_tools/src/main/protobuf/android_deploy_info.proto
     1559  1980-01-01 00:00   embedded_tools/src/main/protobuf/bazel_flags.proto
    20956  1980-01-01 00:00   embedded_tools/src/main/protobuf/build.proto
     3079  1980-01-01 00:00   embedded_tools/src/main/protobuf/builtin.proto
     3874  1980-01-01 00:00   embedded_tools/src/main/protobuf/bundlemerge.proto
     4184  1980-01-01 00:00   embedded_tools/src/main/protobuf/command_line.proto
     5709  1980-01-01 00:00   embedded_tools/src/main/protobuf/command_server.proto
    21254  1980-01-01 00:00   embedded_tools/src/main/protobuf/crosstool_config.proto
     2177  1980-01-01 00:00   embedded_tools/src/main/protobuf/deps.proto
     2035  1980-01-01 00:00   embedded_tools/src/main/protobuf/desugar_deps.proto
     1997  1980-01-01 00:00   embedded_tools/src/main/protobuf/execution_statistics.proto
     6419  1980-01-01 00:00   embedded_tools/src/main/protobuf/extra_actions_base.proto
     8844  1980-01-01 00:00   embedded_tools/src/main/protobuf/invocation_policy.proto
     1325  1980-01-01 00:00   embedded_tools/src/main/protobuf/java_compilation.proto
     1713  1980-01-01 00:00   embedded_tools/src/main/protobuf/option_filters.proto
     2765  1980-01-01 00:00   embedded_tools/src/main/protobuf/plmerge.proto
     5548  1980-01-01 00:00   embedded_tools/src/main/protobuf/remote_execution_log.proto
     4388  1980-01-01 00:00   embedded_tools/src/main/protobuf/spawn.proto
     3441  1980-01-01 00:00   embedded_tools/src/main/protobuf/test_status.proto
     1808  1980-01-01 00:00   embedded_tools/src/main/protobuf/worker_protocol.proto
      625  1980-01-01 00:00   embedded_tools/src/tools/android/java/com/google/devtools/build/android/proto/BUILD
     5012  1980-01-01 00:00   embedded_tools/src/tools/android/java/com/google/devtools/build/android/proto/Configuration.proto
    15647  1980-01-01 00:00   embedded_tools/src/tools/android/java/com/google/devtools/build/android/proto/Resources.proto
     1859  1980-01-01 00:00   embedded_tools/src/tools/android/java/com/google/devtools/build/android/proto/ResourcesInternal.proto
     3008  1980-01-01 00:00   embedded_tools/src/tools/android/java/com/google/devtools/build/android/proto/serialize_format.proto
      722  1980-01-01 00:00   embedded_tools/tools/objc/dummy.proto
      820  1980-01-01 00:00   embedded_tools/tools/objc/protobuf_compiler.py
      684  1980-01-01 00:00   embedded_tools/tools/objc/protobuf_compiler_wrapper.sh
       89  1980-01-01 00:00   embedded_tools/tools/objc/protobuf_support

**Comment by davido** [comment]
@laszlocsomor Thank you for the detailed explanation. I suspected, that `rules_closure` is involved here, thanks for clarifying. And that why I've upgraded `rules_closure` in Gerrit, in: [1]. I'm currently away from my computer, but I'm confused, why with: [1] applied, the above test didn't work.

[1] https://gerrit-review.googlesource.com/c/gerrit/+/218586-

**Comment by laszlocsomor** [comment]
You're welcome. This is all extremely confusing.

IIUC https://gerrit-review.googlesource.com/c/gerrit/+/218586 updates Gerrit's reference of rules_closure to https://github.com/bazelbuild/rules_closure/commit/316e6133888bfc39fb860a4f1a31cfcbae485aef, but that still uses 3.6.1.3: https://github.com/bazelbuild/rules_closure/blob/316e6133888bfc39fb860a4f1a31cfcbae485aef/closure/repositories.bzl#L678

Do you agree?

**Comment by davido** [comment]
> [..] but that still uses 3.6.1.3: https://github.com/bazelbuild/rules_closure/blob/316e6133888bfc39fb860a4f1a31cfcbae485aef/closure/repositories.bzl#L678
>
> Do you agree?

Absolutely! Thanks for finding my mistake. I see what happened. I had two workspaces for rules_closure: one the upstream and another for my fork. I implemented the complete update in upstream clone, and then copied the changed files to my fork to create the PR, and of course I missed closure/repositories.bzl ;-(

I fixed that and have sent the second PR to actually upgrade rules_closure to 3.7.0, here: https://github.com/bazelbuild/rules_closure/pull/342.

@laszlocsomor But still, Bazel needs the protobuf upgrade, to be forward compatible. Otherwise, say this: `--incompatible_disable_deprecated_attr_params` and other incompatible options would be flipped or even removed in recent Bazel releases and that would mean that the most recent Bazel release wouldn't be able to build Bazel@HEAD.

Basically, the same reasoning for upgrading Protobuf in rules_closure to make Gerrit bazel build toolchain forward compatible holds for Bazel itself.

Of course, Bazel team could set up .bazelversion to configure Baselisk to build Bazel@HEAD with some ancient Bazel version, before the incompatible options were flipped and removed ;-)

**Comment by laurentlb** [comment]
> But still, Bazel needs the protobuf upgrade, to be forward compatible.

+1

**Comment by laszlocsomor** [comment]
@davido : You're welcome!

I agree Bazel should update its own protobuf, but @buchgr tells me it's risky. If there's no immediate need to update then I prefer to postpone and close this PR. Sounds reasonable?

**Comment by davido** [comment]
> If there's no immediate need to update then I prefer to postpone and close this PR. Sounds reasonable?

Sounds fine to me. However, the tricky part, as can be seen from upgrade attempt, arose from recent dependencies added to Protobuf library:

1. Bazel skylib (that I have reverted in my upgrade attempt)
2. zlib (that I routed to the Baze'sl own version of zlib in my upgrade attempt, with the complication, that the integration tests didn't work with Bazel's own zlib version, and was changed to consumption of yet another zlib from upstream in: https://github.com/bazelbuild/bazel/pull/7681/commits/c3d06104dab1ba6825a826ab8c1c2a877cdf908e)

If we abandon the attempt to upgrade Protobuf today, the above problem would not be magically solved tomorrow. So that we probably should start to address them in mid/long term.

**Comment by laszlocsomor** [comment]
There's no progress, no immediate need, and nobody working on this PR, so I'm closing it.
